src/Sulu/Bundle/ContentBundle/Controller/NodeController.php

```diff
@@ -574,48 +580,70 @@ public function filterAction(Request $request)
      * saves node with given uuid and data.
      *
      * @param \Symfony\Component\HttpFoundation\Request $request
-     * @param string                                    $uuid
+     * @param string $uuid
      *
-     * @return \Symfony\Component\HttpFoundation\Response
+     * @return Response
+     *
+     * @throws InvalidFormException
+     * @throws MissingParameterException
      */
     public function putAction(Request $request, $uuid)
     {
         $language = $this->getLanguage($request);
-        $webspace = $this->getWebspace($request);
-        $isShadow = $this->getRequestParameter($request, 'shadowOn', false);
-        $shadowBaseLanguage = $this->getRequestParameter($request, 'shadowBaseLanguage', null);
+        $type = $request->get('type', 'page');
 
-        $state = $this->getRequestParameter($request, 'state');
-        $type = $request->query->get('type', 'page');
+        try {
+            $document = $this->getDocumentManager()->find(
+                $uuid,
+                $language,
+                ['type' => $type, 'load_ghost_content' => false]
+            );
+        } catch (DocumentNotFoundException $e) {
+            $e = new EntityNotFoundException(PageDocument::class, $uuid, $e);
 
-        if ($state !== null) {
-            $state = intval($state);
+            return $this->handleView($this->view($e->toArray(), 404));
         }
 
+        $webspace = $this->getWebspace($request);
         $data = $request->request->all();
+        $data['structureType'] = $data['template'];
+        $data['shadowLocaleEnabled'] = $this->getRequestParameter($request, 'shadowOn', false);
+        $data['shadowLocale'] = $this->getRequestParameter($request, 'shadowBaseLanguage', null);
+        $data['workflowStage'] = $this->getRequestParameter($request, 'state');
```

Comment:
Have a look at the `DataNormalizer`.

Previously we passed these request parameters as arguments to the content mapper, which meant that they had not needed to be handled by the `DataNormalizer`. There is however, no reason that the `DataNormalizer` cannot handle this.

You will need to update the `DataNormalizer` to map these fields, and update the `ContentMapper::save()` method to map the arguments to the *legacy* fields (i.e. `shadowOn`, `shadowBaseLanguage`, etc).